packages/theme/src/cli/utilities/theme-uploader.ts

```diff
@@ -32,82 +33,171 @@ export async function uploadTheme(
   const remoteChecksums = rejectGeneratedStaticAssets(checksums)
   const uploadResults: Map<string, Result> = new Map()
   await themeFileSystem.ready()
-  const deleteTasks = await buildDeleteTasks(remoteChecksums, themeFileSystem, options, theme, session)
-  const uploadTasks = await buildUploadTasks(remoteChecksums, themeFileSystem, options, theme, session, uploadResults)
 
-  // The task execution mechanism processes tasks sequentially in the order they are added.
-  await renderTasksToStdErr(deleteTasks)
-  await renderTasksToStdErr(uploadTasks)
+  const {progress: uploadProgress, promise: uploadPromise} = await buildUploadJob(
+    remoteChecksums,
+    themeFileSystem,
+    options,
+    theme,
+    session,
+    uploadResults,
+  )
 
-  reportFailedUploads(uploadResults)
-  return uploadResults
+  const {promise: deletePromise} = await buildDeleteJob(remoteChecksums, themeFileSystem, options, theme, session)
+
+  return {
+    uploadResults,
+    renderThemeSyncProgress: async () => {
+      // The task execution mechanism processes tasks sequentially in the order they are added.
+
+      const getProgress = (params: {current: number; total: number}) =>
+        `[${Math.round((params.current / params.total) * 100)}%]`
+
+      await renderTasksToStdErr(
+        createIntervalTask({
+          promise: uploadPromise,
+          titleGetter: () => `Uploading files to remote theme ${getProgress(uploadProgress)}`,
+          timeout: 1000,
+        }),
+      )
+
+      deletePromise
+        .then(() => {})
+        .catch(() => {
+          renderWarning({headline: 'Failed to delete outdated files from remote theme.'})
+        })
+
+      reportFailedUploads(uploadResults)
+    },
+  }
+}
+
+function createIntervalTask({
+  promise,
+  titleGetter,
+  timeout,
+}: {
+  promise: Promise<unknown>
+  titleGetter: () => string
+  timeout: number
+}) {
+  const tasks: Task[] = []
+
+  const addNextCheck = () => {
+    tasks.push({
+      title: titleGetter(),
+      task: async () => {
+        const result = await Promise.race([
+          promise,
+          new Promise((resolve) => setTimeout(() => resolve('timeout'), timeout)),
+        ])
+
+        if (result === 'timeout') {
+          addNextCheck()
+        }
```

Comment:
This adds a new task when the timeout ends until `promise` is resolved. This way we update the title with a progress %